storage/segment_query.go

```diff
@@ -654,3 +596,44 @@ func collectUnorderedMultiFieldGroupByResults(
 	}
 	return groupByCalcIter.Err()
 }
+
+func newSingleTypeFieldIterator(
+	fieldPath []string,
+	allowedTypes field.ValueTypeSet,
+	queryField indexfield.DocsField,
+	expectedType *field.ValueType,
+) (indexfield.BaseFieldIterator, field.ValueType, error) {
+	if len(allowedTypes) != 1 {
```

Comment:
why not just pass into `field.ValueType` instead?

Reply:
Wouldn't that require we do the logic from line 606 to line 613 in every place before calling `newSingleTypeFieldIterator`? This also validates that if there are multiple types for this field in the segment that error is caught here.